src/query/functions-v2/src/scalars/control.rs

```diff
@@ -173,4 +186,108 @@ pub fn register(registry: &mut FunctionRegistry) {
             ValueRef::Scalar(Some(_)) => Ok(Value::Scalar(true)),
         },
     );
+
+    fn cal_in<T: ArgType>(
+        lhs: ValueRef<T>,
+        rhs: ValueRef<ArrayType<T>>,
+    ) -> Result<Value<BooleanType>, String>
+    where
+        T::Scalar: HashtableKeyable,
+    {
+        // must be scalar in rhs, if rhs contains column, it'll be rewrited to `a = b1 or a = b2 or a = b3` in type_checker
+        let array = rhs.as_scalar().unwrap();
+        let mut set = StackHashSet::<_, 128>::with_capacity(T::column_len(array));
+        for val in T::iter_column(array) {
+            let _ = set.set_insert(T::to_owned_scalar(val));
+        }
+        match lhs {
+            ValueRef::Scalar(c) => Ok(Value::Scalar(set.contains(&T::to_owned_scalar(c)))),
+            ValueRef::Column(col) => {
+                let result = BooleanType::column_from_iter(
+                    T::iter_column(&col).map(|c| set.contains(&T::to_owned_scalar(c))),
+                    &[],
+                );
+                Ok(Value::Column(result))
+            }
+        }
+    }
+
+    for left in ALL_NUMERICS_TYPES {
+        with_number_mapped_type!(|NUM_TYPE| match left {
+            NumberDataType::NUM_TYPE => {
+                registry.register_passthrough_nullable_2_arg::<NumberType<NUM_TYPE>, ArrayType<NumberType<NUM_TYPE>>, BooleanType, _, _>(
+                    "in",
+                    FunctionProperty::default(),
+                    |_, _| None,
+                    |lhs, rhs, _| cal_in(lhs, rhs)
+                );
+            }
+        });
+    }
+
+    registry
+        .register_passthrough_nullable_2_arg::<DateType, ArrayType<DateType>, BooleanType, _, _>(
```

Comment:
do you think that we need to implement `in(GenericType<0>, ArrayType<GenericType<0>>) :: BooleanType` and `in(VariantType, ArrayType<VariantType>) :: BooleanType`?

Reply:
We can add them if we really need them. For now, they are rarely used.